include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -0,0 +1,301 @@
+#ifndef _NBL_BUILTIN_HLSL_CPP_COMPAT_IMPL_INTRINSICS_IMPL_INCLUDED_
+#define _NBL_BUILTIN_HLSL_CPP_COMPAT_IMPL_INTRINSICS_IMPL_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace cpp_compat_intrinsics_impl
+{
+template<typename T>
+struct dot_helper
+{
+	using scalar_type = typename vector_traits<T>::scalar_type;
+
+	static inline scalar_type dot(NBL_CONST_REF_ARG(T) lhs, NBL_CONST_REF_ARG(T) rhs)
+	{
+		static array_get<T, scalar_type> getter;
+		scalar_type retval = getter(lhs, 0) * getter(rhs, 0);
+
+		static const uint32_t ArrayDim = sizeof(T) / sizeof(scalar_type);
+		for (uint32_t i = 1; i < ArrayDim; ++i)
+			retval = retval + getter(lhs, i) * getter(rhs, i);
+
+		return retval;
+	}
+};
+
+#define DEFINE_BUILTIN_VECTOR_SPECIALIZATION(FLOAT_TYPE, RETURN_VALUE)\
+template<uint32_t N>\
+struct dot_helper<vector<FLOAT_TYPE, N> >\
+{\
+	using VectorType = vector<FLOAT_TYPE, N>;\
+	using ScalarType = typename vector_traits<VectorType>::scalar_type;\
+\
+	static inline ScalarType dot(NBL_CONST_REF_ARG(VectorType) lhs, NBL_CONST_REF_ARG(VectorType) rhs)\
+	{\
+		return RETURN_VALUE;\
+	}\
+};\
+
+#ifdef __HLSL_VERSION
+#define BUILTIN_VECTOR_SPECIALIZATION_RET_VAL dot(lhs, rhs)
+#else
+#define BUILTIN_VECTOR_SPECIALIZATION_RET_VAL glm::dot(lhs, rhs)
+#endif
+
+DEFINE_BUILTIN_VECTOR_SPECIALIZATION(float16_t, BUILTIN_VECTOR_SPECIALIZATION_RET_VAL)
+DEFINE_BUILTIN_VECTOR_SPECIALIZATION(float32_t, BUILTIN_VECTOR_SPECIALIZATION_RET_VAL)
+DEFINE_BUILTIN_VECTOR_SPECIALIZATION(float64_t, BUILTIN_VECTOR_SPECIALIZATION_RET_VAL)
+
+#undef BUILTIN_VECTOR_SPECIALIZATION_RET_VAL
+#undef DEFINE_BUILTIN_VECTOR_SPECIALIZATION
+
+#ifdef __HLSL_VERSION
+template<typename Integer>
+struct find_msb_helper;
+#else
+// legacy code wouldn't work without it
+template<typename Integer>
+struct find_msb_helper
+{
+	static int findMSB(NBL_CONST_REF_ARG(Integer) val)
+	{
+		if (is_signed_v<Integer>)
+		{
+			// GLM accepts only integer types, so idea is to cast input to integer type
+			using as_int = typename integer_of_size<sizeof(Integer)>::type;
+			const as_int valAsInt = reinterpret_cast<const as_int&>(val);
+			return glm::findMSB(valAsInt);
+		}
+		else
+		{
+			// GLM accepts only integer types, so idea is to cast input to integer type
+			using as_uint = typename unsigned_integer_of_size<sizeof(Integer)>::type;
+			const as_uint valAsUnsignedInt = reinterpret_cast<const as_uint&>(val);
+			return glm::findMSB(valAsUnsignedInt);
+		}
```

Comment:
when were we using `findMSB` on floats or or something similar?

Reply:
also you have a bunch of specializations later which call glm, what is this default spec for? other intger sizes than 32 ?

Reply:
> when were we using `findMSB` on floats or or something similar?

mostly `core::bitflag`, which you can't static_cast to underlying type